FunctionalSpec.md

```diff
@@ -4,24 +4,74 @@
+	- Note: we are not asking you to purchase a Particle device to actually run this code.
+- Serial communications will be sent & received via [Serial](https://docs.particle.io/cards/firmware/serial/serial/) device using **Opti Vending Machine Message Protocol** defined below.
+- Does not depend on any external web services.
 - Does not require any persistent data storage.
-- Only manages 1 coffee order at once. You can assume whatever physical apparatus is connected to the other end of this service prevents more than 1 user from using it at once.
+- Only manages 1 coffee order at once. One order may consist of multiple coffees of different sizes. You can assume whatever physical apparatus is connected to the other end of this service prevents more than 1 user from using it at once.
 
 ## Functional Requirements
+
 - Order coffee in 3 sizes (small for $1.75, medium for $2.00, large for $2.25).
 - Allow multiple coffee orders per payment transaction, up to 5 coffees.
 - Allow adding money in standard monetary increments only (from $0.05 to $20).
 - Allow user to specify the end of a payment transaction and "Dispense" coffee and change.
-	- “Dispense” coffee when an order is completed if adequate payment has been provided and display money remaining in payment transaction. The developer can assume that the physical implications of this dispense operation always succeed and do not need to be verified - in this context, "dispense" means complete the order.
-- Update the user via events sent to Particle.publish() regarding the current state of the order. Give warning if the user attempts to "Dispense" without adding adequate payment.
-- Document the different HTTP calls that drive this system.
+- “Dispense” coffee when an order is completed if adequate payment has been provided and display money remaining in payment transaction. The developer can assume that the physical implications of this dispense operation always succeed and do not need to be verified - in this context, "dispense" means complete the order.
+- Insertion of money is communicated to device by serial message
+- Device sends updates to user via events sent via Serial communications
+- At start up display message of the day on LCD
+
+## Implementation Requirements
+
+- Small Button is GPIO PIN A0
+- Medium Button is GPIO PIN A1
+- Large Button is GPI0 PIN A2
+- Dispense/Cancel Button GPIO Pin A3
+   - Dispense will be triggered by press & release
+   - Cancel will be triggered by press & hold > 2 seconds
+- Message of the day value to write to LCD using `setLCD` will be evaluated using last received `msgOfDay`
+  template value and replacing each occurrence of `%KEY%` with the corresponding value for that
+  `KEY` where `KEY` is one of the supported keys excluding `msgOfDay` and `setLCD`.  If `KEY` is
+  an unknown key or `msgOfDay` or `setLCD`, then it will not be replaced.

```

Comment:
A few questions relating to this feature:
- Should we clarify that `msgOfDay` will be sent at startup and then randomly thereafter? Does the device ever have to run without a last received `msgOfDay`?
- Do all of the `KEY`s have defined values at all times, i.e. even when the device is not in the middle of handling an order?
- Also, although we have defined serial and GPIO, it is not explicit IMO what to do to "write to the LCD." Should we define that either in the Assumptions and Constraints section where we talk about serial or above here where we define the GPIOs?

Reply:
Some of the `KEY`s imply there are associated state variables as well.  And yes, I think it would be good to define the initialized values for this state:

Initialization State:
* `msgDay = '' //empty ASCII string`
* `order = {0x0, 0x0, 0x0} // zero small, zero medium, zero large coffees`
* `curFunds = 0`

I agree it is not specific about write to the LCD means, but my hope is that they can infer `setLCD` message is the means to write to the LCD.  We could state this explicitly, but I think it may also be a good test of whether they can infer this on their own because often documentation has ambiguity like this.  I could go either way about stating it explicitly or leaving it as is.

Reply:
Actually, it already says
> Message of the day value to write to LCD using `setLCD`...

Reply:
I clarified bullet on how to write to LCD.  And added initialized state values.  Does this look better?

Reply:
I see your point about the message implying it, and I think this tweak makes it clear enough, thank you. +1 on the initialization as well.